lib/types/array.js

```diff
@@ -664,20 +664,26 @@ internals.fillMissedErrors = function (schema, errors, requireds, value, state,
 internals.checkRemainingOrdereds = function (schema, errors, ordereds, value, state, prefs) {
 
     const requiredOrdereds = [];
+    const orderedsWithDefault = [];
 
     for (const ordered of ordereds) {
         if (ordered._flags.presence === 'required') {
             requiredOrdereds.push(ordered);
         }
 
-        if (ordered._flags.hasOwnProperty('default')) {
-            value.push(ordered._flags.default);
+        const res = ordered.$_validate(undefined,state,prefs);
+        if (res.value !== undefined) {
+            orderedsWithDefault.push(res.value);
```

Comment:
Wouldn't this mean 
```js
Joi.array().ordered(Joi.string().default('a'), Joi.string(), Joi.string().default('b')
```
will return `['a', 'b']` instead of `['a', undefined, 'b']`? The order of 'b' is no longer retained

Reply:
Looks alright. Can you please add tests for more complex schemas and see how it goes?

Reply:
Yes it wouldn't be retained at all, and this's kinda expected i think if there's a **not** `required` field in the middle, don't you agree?
Sure i will add a test confirming this.

Reply:
I think the correct behaviour here should be the latter because the order is what the schema describes.

Reply:
We also need to consider trailing sparsed items. 
```js
Joi.array().ordered(Joi.string().default('a'), Joi.number());
```
What should be the behaviour here? `['a', undefined]` or `['a']`?
This might not make a big difference, but `.length` would return different values here.

Reply:
The correct behaviour here, imo would be to discard any trailing item. We could achieve this with a reversed loop, discarding any `undefined` until an actual value is found.

Reply:
@iifawzi

> 
> 
> Yes it wouldn't be retained at all, and this's kinda expected i think if there's a **not** `required` field in the middle, don't you agree?
> Sure i will add a test confirming this.

I think the order of items is more important here. After all, the method being used is `ordered`, and it should attempt to match the schema as much as possible.

Reply:
> What should be the behaviour here? ['a', undefined] or ['a']?

hmm, imo? it should be `['a']` :d, but you're right, since it's `ordered` it must match the schema as much as possible, so should we git rid of the `undefined` check at all? or somehow letting the user choose the behavior he wants? (filling not inserted items with `undefined` or not)